File: web/src/shared/lib/search/__tests__/build-filter-by.test.ts
/**
 * build-filter-by.test.ts is the parity test for the TS filter
 * builder. The same scenarios are tested in the Go side
 * (`backend/internal/app/search/filter_builder_test.go`); both
 * tests must produce identical filter_by strings.
 */

import { describe, expect, it } from "vitest"
import { buildFilterBy, type SearchFilterInput } from "../build-filter-by"

describe("buildFilterBy", () => {
  it("returns empty string for empty input", () => {
    expect(buildFilterBy({})).toBe("")
  })

  it("emits availability_status clause for non-empty list", () => {
    expect(buildFilterBy({ availabilityStatus: ["available_now"] })).toBe(
      "availability_status:[available_now]",
    )
    expect(
      buildFilterBy({ availabilityStatus: ["available_now", "available_soon"] }),
    ).toBe("availability_status:[available_now,available_soon]")
  })

  it("dedupes and trims string slices", () => {
    expect(
      buildFilterBy({
        availabilityStatus: [" available_now ", "available_now", "  "],
      }),
    ).toBe("availability_status:[available_now]")
  })

  it("emits pricing range clause", () => {
    expect(buildFilterBy({ pricingMin: 50000, pricingMax: 150000 })).toBe(
      "pricing_min_amount:>=50000 && pricing_min_amount:<=150000",
    )
    expect(buildFilterBy({ pricingMin: 50000 })).toBe("pricing_min_amount:>=50000")
    expect(buildFilterBy({ pricingMax: 150000 })).toBe("pricing_min_amount:<=150000")
  })

  it("emits case-insensitive city clause with backticks", () => {
    expect(buildFilterBy({ city: "Paris" })).toBe("city:`Paris`")
    expect(buildFilterBy({ city: "New York" })).toBe("city:`New York`")
    expect(buildFilterBy({ city: "  Lyon  " })).toBe("city:`Lyon`")
    expect(buildFilterBy({ city: "" })).toBe("")
    expect(buildFilterBy({ city: "   " })).toBe("")
  })

  it("preserves country code casing (match is case-insensitive)", () => {
    expect(buildFilterBy({ countryCode: "FR" })).toBe("country_code:FR")
    expect(buildFilterBy({ countryCode: "fr" })).toBe("country_code:fr")
    expect(buildFilterBy({ countryCode: "" })).toBe("")
  })

  it("emits geo clause when lat/lng/radius all set", () => {
    expect(
      buildFilterBy({ geoLat: 48.8566, geoLng: 2.3522, geoRadiusKm: 25 }),
    ).toBe("location:(48.8566,2.3522,25 km)")
  })

  it("drops geo clause when any field is missing", () => {
    expect(buildFilterBy({ geoLat: 48.8566 })).toBe("")
    expect(buildFilterBy({ geoLat: 48.8566, geoLng: 2.3522 })).toBe("")
    expect(
      buildFilterBy({ geoLat: 48.8566, geoLng: 2.3522, geoRadiusKm: 0 }),
    ).toBe("")
  })

  it("emits language / expertise / skill list clauses", () => {
    expect(buildFilterBy({ languages: ["fr", "en"] })).toBe(
      "languages_professional:[fr,en]",
    )
    expect(buildFilterBy({ expertiseDomains: ["dev", "design"] })).toBe(
      "expertise_domains:[dev,design]",
    )
    expect(buildFilterBy({ skills: ["react", "go"] })).toBe("skills:[react,go]")
  })

  it("emits rating clause with smart number formatting", () => {
    expect(buildFilterBy({ ratingMin: 4 })).toBe("rating_average:>=4")
    expect(buildFilterBy({ ratingMin: 4.5 })).toBe("rating_average:>=4.5")
    expect(buildFilterBy({ ratingMin: 0 })).toBe("")
    expect(buildFilterBy({ ratingMin: -1 })).toBe("")
  })

  it("emits work_mode clause", () => {
    expect(buildFilterBy({ workMode: ["remote", "hybrid"] })).toBe(
      "work_mode:[remote,hybrid]",
    )
  })

  it("emits boolean toggles", () => {
    expect(buildFilterBy({ isVerified: true })).toBe("is_verified:=true")
    expect(buildFilterBy({ isVerified: false })).toBe("is_verified:=false")
    expect(buildFilterBy({ isTopRated: true })).toBe("is_top_rated:=true")
    expect(buildFilterBy({ negotiable: true })).toBe("pricing_negotiable:=true")
  })

  it("preserves stable order across all clauses (parity with Go)", () => {
    const got = buildFilterBy({
      availabilityStatus: ["available_now"],
      pricingMin: 40000,
      pricingMax: 120000,
      city: "Paris",
      countryCode: "FR",
      languages: ["fr", "en"],
      skills: ["react"],
      ratingMin: 4,
      workMode: ["remote"],
      isVerified: true,
      isTopRated: true,
    })
    expect(got).toBe(
      "availability_status:[available_now]" +
        " && pricing_min_amount:>=40000 && pricing_min_amount:<=120000" +
        " && city:`Paris`" +
        " && country_code:FR" +
        " && languages_professional:[fr,en]" +
        " && skills:[react]" +
        " && rating_average:>=4" +
        " && work_mode:[remote]" +
        " && is_verified:=true" +
        " && is_top_rated:=true",
    )
  })

  it("is deterministic for repeated calls", () => {
    const input: SearchFilterInput = {
      languages: ["fr"],
      skills: ["go", "react"],
    }
    const a = buildFilterBy(input)
    const b = buildFilterBy(input)
    expect(a).toBe(b)
  })
})

File: mobile/test/shared/search/build_filter_by_test.dart
// build_filter_by_test.dart — parity + unit test for the mobile
// Typesense filter_by builder.
//
// The expected strings in the PARITY table below are copied from the
// corresponding Go tests in
// `backend/internal/app/search/filter_builder_test.go` and from the
// TypeScript tests in
// `web/src/shared/lib/search/__tests__/build-filter-by.test.ts`.
// If any of them changes, a matching change must happen here —
// otherwise the mobile filter string will silently diverge from
// what the backend expects.
//
// Parity = byte-for-byte equality, not semantic equivalence.

import 'package:flutter_test/flutter_test.dart';
import 'package:marketplace_mobile/shared/search/build_filter_by.dart';
import 'package:marketplace_mobile/shared/search/search_filters.dart';

void main() {
  group('buildFilterBy parity', () {
    test('empty input returns empty string', () {
      expect(buildFilterBy(const SearchFilterInput()), '');
    });

    test('availability single value', () {
      expect(
        buildFilterBy(
          const SearchFilterInput(availabilityStatus: ['now']),
        ),
        'availability_status:[now]',
      );
    });

    test('availability multi value preserves order', () {
      expect(
        buildFilterBy(
          const SearchFilterInput(availabilityStatus: ['now', 'soon']),
        ),
        'availability_status:[now,soon]',
      );
    });

    test('availability dedupes', () {
      expect(
        buildFilterBy(
          const SearchFilterInput(
            availabilityStatus: ['now', ' now ', 'now', 'soon'],
          ),
        ),
        'availability_status:[now,soon]',
      );
    });

    test('pricing min only', () {
      expect(
        buildFilterBy(const SearchFilterInput(pricingMin: 500)),
        'pricing_min_amount:>=500',
      );
    });

    test('pricing max only', () {
      expect(
        buildFilterBy(const SearchFilterInput(pricingMax: 1200)),
        'pricing_min_amount:<=1200',
      );
    });

    test('pricing min + max joined with &&', () {
      expect(
        buildFilterBy(const SearchFilterInput(pricingMin: 500, pricingMax: 1200)),
        'pricing_min_amount:>=500 && pricing_min_amount:<=1200',
      );
    });

    test('city backtick-escaped', () {
      expect(
        buildFilterBy(const SearchFilterInput(city: 'New York')),
        'city:`New York`',
      );
    });

    test('city trimmed', () {
      expect(
        buildFilterBy(const SearchFilterInput(city: '  Paris  ')),
        'city:`Paris`',
      );
    });

    test('countryCode preserves casing', () {
      expect(
        buildFilterBy(const SearchFilterInput(countryCode: 'FR')),
        'country_code:FR',
      );
    });

    test('geo clause emits lat,lng,N km', () {
      expect(
        buildFilterBy(
          const SearchFilterInput(
            geoLat: 48.8566,
            geoLng: 2.3522,
            geoRadiusKm: 25,
          ),
        ),
        'location:(48.8566,2.3522,25 km)',
      );
    });

    test('geo clause drops when any coordinate missing', () {
      expect(
        buildFilterBy(
          const SearchFilterInput(geoLat: 48.8566, geoRadiusKm: 25),
        ),
        '',
      );
    });

    test('geo clause drops when radius is zero', () {
      expect(
        buildFilterBy(
          const SearchFilterInput(
            geoLat: 48.8566,
            geoLng: 2.3522,
            geoRadiusKm: 0,
          ),
        ),
        '',
      );
    });

    test('languages OR', () {
      expect(
        buildFilterBy(const SearchFilterInput(languages: ['fr', 'en'])),
        'languages_professional:[fr,en]',
      );
    });

    test('expertise OR', () {
      expect(
        buildFilterBy(
          const SearchFilterInput(expertiseDomains: ['development', 'design']),
        ),
        'expertise_domains:[development,design]',
      );
    });

    test('skills OR preserves order', () {
      expect(
        buildFilterBy(
          const SearchFilterInput(skills: ['React', 'Go', 'TypeScript']),
        ),
        'skills:[React,Go,TypeScript]',
      );
    });

    test('rating clause', () {
      expect(
        buildFilterBy(const SearchFilterInput(ratingMin: 4.5)),
        'rating_average:>=4.5',
      );
    });

    test('rating clause zero dropped', () {
      expect(
        buildFilterBy(const SearchFilterInput(ratingMin: 0)),
        '',
      );
    });

    test('work mode list', () {
      expect(
        buildFilterBy(
          const SearchFilterInput(workMode: ['remote', 'hybrid']),
        ),
        'work_mode:[remote,hybrid]',
      );
    });

    test('booleans use := operator', () {
      expect(
        buildFilterBy(const SearchFilterInput(isVerified: true)),
        'is_verified:=true',
      );
      expect(
        buildFilterBy(const SearchFilterInput(isTopRated: false)),
        'is_top_rated:=false',
      );
      expect(
        buildFilterBy(const SearchFilterInput(negotiable: true)),
        'pricing_negotiable:=true',
      );
    });

    test('full payload order matches backend (parity anchor)', () {
      // This is THE parity anchor — a lived-in filter payload that
      // exercises every clause. If this test fails, filter_by has
      // drifted from the Go + TS counterparts.
      final input = const SearchFilterInput(
        availabilityStatus: ['now'],
        pricingMin: 500,
        pricingMax: 1500,
        city: 'Paris',
        countryCode: 'FR',
        geoLat: 48.8566,
        geoLng: 2.3522,
        geoRadiusKm: 25,
        languages: ['fr', 'en'],
        expertiseDomains: ['development'],
        skills: ['React', 'Go'],
        ratingMin: 4,
        workMode: ['remote'],
        isVerified: true,
        isTopRated: true,
        negotiable: false,
      );
      expect(
        buildFilterBy(input),
        'availability_status:[now] && '
        'pricing_min_amount:>=500 && pricing_min_amount:<=1500 && '
        'city:`Paris` && country_code:FR && '
        'location:(48.8566,2.3522,25 km) && '
        'languages_professional:[fr,en] && '
        'expertise_domains:[development] && '
        'skills:[React,Go] && '
        'rating_average:>=4 && '
        'work_mode:[remote] && '
        'is_verified:=true && '
        'is_top_rated:=true && '
        'pricing_negotiable:=false',
      );
    });
  });

  group('SearchFilterInput.fromMap', () {
    test('accepts the filtersToInput payload shape', () {
      final mobileFilters = const MobileSearchFilters(
        availability: MobileAvailabilityFilter.now,
        priceMin: 500,
        priceMax: 1500,
        city: 'Paris',
        countryCode: 'FR',
        radiusKm: 25,
        languages: <String>{'fr', 'en'},
        expertise: <String>{'development'},
        skills: <String>['React'],
        minRating: 4,
        workModes: <MobileWorkMode>{MobileWorkMode.remote},
      );
      final input = SearchFilterInput.fromMap(filtersToInput(mobileFilters));
      final out = buildFilterBy(input);
      expect(
        out,
        'availability_status:[now] && '
        'pricing_min_amount:>=500 && pricing_min_amount:<=1500 && '
        'city:`Paris` && country_code:FR && '
        'languages_professional:[fr,en] && '
        'expertise_domains:[development] && '
        'skills:[React] && '
        'rating_average:>=4 && '
        'work_mode:[remote]',
      );
    });

    test('empty filters yield empty filter_by', () {
      final input = SearchFilterInput.fromMap(
        filtersToInput(kEmptyMobileSearchFilters),
      );
      expect(buildFilterBy(input), '');
    });

    test('radius is dropped when no city/country is set', () {
      final f = const MobileSearchFilters(radiusKm: 25);
      expect(filtersToInput(f).containsKey('geoRadiusKm'), isFalse);
    });

    test('radius is kept when city is set', () {
      final f = const MobileSearchFilters(city: 'Paris', radiusKm: 25);
      expect(filtersToInput(f)['geoRadiusKm'], 25);
    });

    test('unknown keys ignored (forward compat)', () {
      final map = <String, Object?>{'unexpected_key': 'value', 'city': 'Paris'};
      final input = SearchFilterInput.fromMap(map);
      expect(input.city, 'Paris');
    });
  });
}
